feat(telemetry): instrument rafiki (#2299)
interledger/rafiki: packages/documentation/src/content/docs/telemetry/privacy.md

---
title: Privacy
---

> Rafiki telemetry is designed with a strong emphasis on privacy. The system anonymizes user data and refrains from collecting identifiable information. Since transactions can originate from any user to a Rafiki instance, the privacy measures are implemented at the Rafiki instance level in the network. This means that at the individual level, the data is already anonymous as single Rafiki instances service transactions for multiple users.

Reviewer: ```suggestion
Rafiki telemetry is designed with a strong emphasis on privacy. The system anonymizes user data and refrains from collecting identifiable information. Since transactions can originate from any user to a Rafiki instance, the privacy measures are implemented directly at the source (each Rafiki instance). This means that at the individual level, the data is already anonymous as single Rafiki instances service transactions for multiple users.
```

To achieve Local Differential Privacy (LDP), noise is selected from the Laplacian distribution and added to the rounded values. The noise is generated based on a privacy parameter, which is calculated using the sensitivity of the function.

The sensitivity of a function in differential privacy is the maximum amount that any single observation can change the output of the function. In this case, the sensitivity is considered to be the maximum of the rounded value and the bucket size.

The privacy parameter is computed as one-tenth of the sensitivity. This parameter controls the trade-off between privacy and utility: a smaller privacy parameter means more privacy but less utility, and a larger privacy parameter means less privacy but more utility.

The noise, selected from the Laplacian distribution, is then generated using this privacy parameter and added to the rounded value. If the resulting value is zero, it is set to half the bucket size to ensure that the noise does not completely obscure the transaction value.

## Currency Conversion

> Another factor that obscures sensitive data is currency conversion. In cross-currency transactions, we use exchange rates that are not traced back. This introduces an additional layer of noise and further protects the privacy of the transactions.

Author: Please have a look at the additional context provided there.

> Another factor that obscures sensitive data is currency conversion. In cross-currency transactions, we use exchange rates that are not traced back. The rates are provided by the integrating [ASE](/reference/glossary#account-servicing-entity) at runtime and are not persisted. If the needed rates are not provided, external API's for exchange rates are used. The obtained exchange rates are overwritten frequently in this case, with no versioning or history access. This introduces an additional layer of noise and further protects the privacy of the transactions.

Reviewer: maybe we can re-word "traced back" better? something like, "since exchange rates are provided by the ASE internally, they cannot be correlated to an individual transaction"

> Another factor that obscures sensitive data is currency conversion. In cross-currency transactions, we use exchange rates that are not traced back. The rates are provided by the integrating [ASE](/reference/glossary#account-servicing-entity) at runtime and are not persisted. If the needed rates are not provided, external API's for exchange rates are used. The obtained exchange rates are overwritten frequently in this case, with no versioning or history access. This introduces an additional layer of noise and further protects the privacy of the transactions.

Reviewer: ```suggestion
Another factor that obscures sensitive data is currency conversion. In cross-currency transactions, we use exchange rates that are not traced back. The rates are provided by the integrating [ASE](/reference/glossary#account-servicing-entity) at runtime and are not persisted. If the necessary rates are not provided or not available from the ASE, an external API for exchange rates are used. The obtained exchange rates are overwritten frequently in this case, with no versioning or history access. This introduces an additional layer of noise and further protects the privacy of the transactions.
```

## References

Rafiki's telemetry solution is a combination of techniques described in various white papers on privacy-preserving data collection. For more information, you can refer to the following papers:

- [Local Differential Privacy for Human-Centered Computing](https://proceedings.neurips.cc/paper_files/paper/2017/file/253614bbac999b38b5b60cae531c4969-Paper.pdf)
- [Collecting Telemetry Data Privately](https://www.microsoft.com/en-us/research/blog/collecting-telemetry-data-privately/)
- [Collecting Telemetry Data Privately - NeurIPS Publication](https://proceedings.neurips.cc/paper_files/paper/2017/file/253614bbac999b38b5b60cae531c4969-Paper.pdf) by Bolin Ding, Janardhan Kulkarni, Sergey Yekhanin from Microsoft Research.
- [RAPPOR: Randomized Aggregatable Privacy-Preserving Ordinal Response](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/42852.pdf)

# Experimental Transaction Values when using the Algorithm

The following table shows the values in the algorithm when running transactions for different amounts. The raw value increases as you go down the rows of the table.

> (all values are in scale 4)

Reviewer: is " (all values are in scale 4)" necessary? if so, maybe it's worth expanding on this just a bit

Author: As i mentioned before, I would argue that scale 2 is better for telemetry than scale 4.
But scale 4 came down as a requirement for some reason (which I don't know)

Author: by:
1. enforcing privacy (rounding + noise addition)
2. using promql operators that change monotonic increasing counters into rate-of-change graphs

We are already disregarding complete accuracy.

Reviewer: I just meant more so in the context of the example table. We don't need to provide the scale necessarily, since it doesn't affect the values in the example table is all.
